core/templates/dev/head/pages/exploration_player/PlaythroughService.js

```diff
@@ -0,0 +1,226 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service for recording and scrutinizing playthroughs.
+ */
+
+oppia.constant(
+  'STORE_PLAYTHROUGH_URL',
+  '/explorehandler/store_playthrough/<exploration_id>');
+
+oppia.factory('PlaythroughService', [
+  '$http', 'PAGE_CONTEXT', 'PlaythroughObjectFactory',
+  'LearnerActionObjectFactory', 'STORE_PLAYTHROUGH_URL',
+  'UrlInterpolationService', 'StopwatchObjectFactory',
+  'CURRENT_ISSUE_SCHEMA_VERSION', 'ISSUE_TYPE_EARLY_QUIT',
+  'ISSUE_TYPE_MULTIPLE_INCORRECT_SUBMISSIONS',
+  'ISSUE_TYPE_CYCLIC_STATE_TRANSITIONS', 'EARLY_QUIT_THRESHOLD_IN_SECS',
+  'NUM_INCORRECT_ANSWERS_THRESHOLD', 'NUM_REPEATED_CYCLES_THRESHOLD',
+  'CURRENT_ACTION_SCHEMA_VERSION', 'ACTION_TYPE_EXPLORATION_START',
+  'ACTION_TYPE_ANSWER_SUBMIT', 'ACTION_TYPE_EXPLORATION_QUIT',
+  function(
+      $http, PAGE_CONTEXT, PlaythroughObjectFactory, LearnerActionObjectFactory,
+      STORE_PLAYTHROUGH_URL, UrlInterpolationService, StopwatchObjectFactory,
+      CURRENT_ISSUE_SCHEMA_VERSION, ISSUE_TYPE_EARLY_QUIT,
+      ISSUE_TYPE_MULTIPLE_INCORRECT_SUBMISSIONS,
+      ISSUE_TYPE_CYCLIC_STATE_TRANSITIONS, EARLY_QUIT_THRESHOLD_IN_SECS,
+      NUM_INCORRECT_ANSWERS_THRESHOLD, NUM_REPEATED_CYCLES_THRESHOLD,
+      CURRENT_ACTION_SCHEMA_VERSION, ACTION_TYPE_EXPLORATION_START,
+      ACTION_TYPE_ANSWER_SUBMIT, ACTION_TYPE_EXPLORATION_QUIT) {
+    var playthrough = null;
+    var expStopwatch = null;
+
+    var multipleIncorrectStateName = {};
+
+    var cycleIdentifier = {};
+    var visitedStates = [];
+
+    var analyzePlaythrough = function() {
+      var timeSpentInExpInSecs = expStopwatch.getTimeInSecs();
+      // Check for multiple incorrect submissions issue.
+      if (multipleIncorrectStateName.num_times_incorrect >= (
+        NUM_INCORRECT_ANSWERS_THRESHOLD)) {
+        playthrough.issueType = ISSUE_TYPE_MULTIPLE_INCORRECT_SUBMISSIONS;
+        playthrough.issueCustomizationArgs = {
+          state_name: {
+            value: multipleIncorrectStateName.state_name
+          },
+          num_times_answered_incorrectly: {
+            value: multipleIncorrectStateName.num_times_incorrect
+          }
+        };
+        return;
+      }
+
+      // Check for cyclic state transitions issue.
+      if (cycleIdentifier.num_cycles >= NUM_REPEATED_CYCLES_THRESHOLD) {
+        playthrough.issueType = ISSUE_TYPE_CYCLIC_STATE_TRANSITIONS;
+        playthrough.issueCustomizationArgs = {
+          state_names: {
+            value: cycleIdentifier.cycle.split(',')
+          }
+        };
+        return;
+      }
+
+      // Check for early quit issue.
+      if (timeSpentInExpInSecs < EARLY_QUIT_THRESHOLD_IN_SECS) {
```

Comment:
refactor this into a function: `isEarlyQuitIssue`.

Reply:
Done.